src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateCopy = {
+    ...state,
+  };
+  const previousVersions = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        for (const key in actions[i].extraData) {
+          stateCopy[key] = actions[i].extraData[key];
+        }
+        break;
+      case 'removeProperties':
+        for (let x = 0; x < actions[i].keysToRemove.length; x++) {
```

Comment:
better to use `for of` loop or forEach method, as we need to iterate through full array

Reply:
> better to use `for of` loop or forEach method, as we need to iterate through full array

why isn't there anything about the forEach method in the lessons about objects, I don't seem to have missed anything